Xiaomi Smart Air Purifier 4 Pro and Air Purifier 4 Lite

jedziemyjedziemy opened the pull request:
Add Xiaomi Air Purifier 4 and Xiaomi Air Purifier 4 Lite. I set up WIP because python-miio on github already has these two devices, but on pypi it has not been updated yet.

Fix #275

Commit:
add airpurifier 4 pro and 4 lite

Commit:
add missed attributes

RafZie commented:
is it possible to add also Xiaomi Smart Air Purifier 4 Pro (zhimi.airp.vb4)?

redding1 commented:
+1 for zhimi.airp.vb4
https://github.com/rytilahti/python-miio/issues/1324
https://github.com/rytilahti/python-miio/pull/1399

syssi commented:
@jedziemyjedziemy thanks for your contribution! Is this pull request complete and well tested?

gabriellferreira commented:
Hey folks, if you want help for testing on `zhimi.airp.mb5` just let me know and help me with some instructions :)

MrFox131 commented:
> Hey folks, if you want help for testing on `zhimi.airp.mb5` just let me know and help me with some instructions :)

i have an mb5a. Relevant? @gabriellferreira 

Just setted up home assitant on my laptop with docker compose for testing purposes

MrFox131 commented:
> @jedziemyjedziemy thanks for your contribution! Is this pull request complete and well tested?
@syssi 
I think I can help with testing on Air Purifier 4, but need a bit of instrutions. I am new to home assistant

MrFox131 commented:
Also can fix #301 #310 #312

MrFox131 commented:
Rename PR please. There is no support for Xiaomi Air Putifier 4, just for pro an lite.
Pro with id zhimi.airp.mb5a is not supported here

IATkachenko commented:
Is any updates for this PR?
```
2023-04-21 09:36:08.191 ERROR (MainThread) [custom_components.xiaomi_miio_airpurifier.fan] Unsupported device found! Please create an issue at https://github.com/syssi/xiaomi_airpurifier/issues and provide the following data: zhimi.airp.rmb1
```
Three other Xiaomi integrations in HA is allow to add this purifier, but does not allow control fan speed via `favorite_level`.

adriannowak commented:
My device has `deerma.humidifier.jsq2w` that is not supported even with this PR. Can we add it?

Kaan88 commented:
I just tested this pr with my air purifier4 pro and it works.

caniko commented:
Why isn't this merged?

Andurilll commented:
> I just tested this pr with my air purifier4 pro and it works.

Can you help me how i can do this for my pr? https://github.com/home-assistant/core/pull/140502
How can i test this locally? 
(i already setup a homeassistant for testing only in proxmox as lxc container but am not shure how to verify if i even test the changes)

syssi committed:
Fix dead code: move 4_PRO/4_LITE elif blocks before PURIFIER_MIOT check

Both models were added to PURIFIER_MIOT, so the model-specific elif
branches at the end of the chain were unreachable and the devices always
fell through to FEATURE_FLAGS_AIRPURIFIER_3 / AVAILABLE_ATTRIBUTES_AIRPURIFIER_3.

syssi committed:
Fix KeyError: align operation mode names with AirpurifierMiotOperationMode enum

"Sleep" and "Manual" are not members of the OperationMode enum in
python-miio (Auto=0, Silent=1, Favorite=2, Fan=3). Selecting either
from the HA frontend raised KeyError in async_set_preset_mode.
Use the enum member names that map to the same integer values on device.

syssi committed:
Fix AttributeError: remove device-display-unit from available attributes

"device-display-unit" contains hyphens and is not a valid Python
identifier. getattr(state, "device-display-unit") raises AttributeError
which is not caught by the DeviceException handler in async_update,
crashing the entity's update loop. python-miio exposes no property for
this field; drop the attribute until a proper accessor exists.

syssi committed:
Fix feature flag: remove FEATURE_SET_FAN_LEVEL from 4 Lite (rmb1)

The MiOT mapping for zhimi.airp.rmb1 (_MAPPING_RMB1 in python-miio)
has no fan_level property. Enabling FEATURE_SET_FAN_LEVEL caused
async_set_fan_level to call set_property("fan_level", ...) on a
property the device does not expose, resulting in a DeviceException.

syssi committed:
Fix pre-commit: ruff format
